### PR TITLE
feat: funnel chart

### PR DESCRIPTION
**Pull request by magnew**
Closes: #2594 

### Description:

Adds a funnel chart viz (from eCharts). This is a draft for now but it's not far from being ready. At the moment you can pick which column in the data to use for the steps. Customers requested also to be able to use a row. We can think about other options (colors?). 

Here's what it looks like for now.

Kapture 2024-06-17 at 19 51 26

### Reviewer actions

- [ ] I have manually tested the changes in the preview environment
- [ ] I have reviewed the code
- [ ] I understand that "request changes" will block this PR from merging

### CODE REVIEW

**Comment by hamzahc1** [comment]
Colours would definitely help here - a few other suggestions: 
- Hide or show label?
- Change label position? (inside - left - top)

**Comment by magnew** [comment]
> Colours would definitely help here - a few other suggestions:
> 
> * Hide or show label?
> * Change label position? (inside - left - top)

@hamzahc1 

Totally doable. What do you think the priority for colors is? Being able to specify a color for each step? Or different auto coloring?

I can add those label options.

**Comment by hamzahc1** [comment]
Likely pulling from the colour scheme initially if it’s set. Otherwise
changing each step manually has the most control. At some point better auto
colouring but feels like a nice to have.

**Comment by magnew** [comment]
@IrakliJani and/or @joaoviana, this is pretty much ready for review. I've fixed the blockers Katie had and I will follow up with the config options Hamzah is requesting. 

I added this dashboard to the render deployment to show the two different data formats: https://lightdash-pr-10445.onrender.com/projects/3675b69e-8324-4110-bdca-059031aa8da3/dashboards/c8333ddc-206f-4b9a-8960-06d2794922d8/view

Let me know if you have any questions!

File: packages/frontend/src/components/FunnelChart/FunnelChartContextMenu.tsx
Comment (magnew) [inline packages/frontend/src/components/FunnelChart/FunnelChartContextMenu.tsx:1]: This is almost entirely the same as PieChartContextMenu. I want to unify them, but I want to do that after this review so I'm not editting the pie chart here.

**Comment by hamzahc1** [comment]
Not relevant for this PR but when using columns as the steps - not being able to easily re-order the columns in the table or in the chart config could be quite annoying

**Comment by joaoviana** [comment]
Early feedback (you can record these for future tickets, non-blocking)

- When there's no metric selected I would like to see an empty state
image

- When I have data orientation as `columns` and I hover on the chart and click `copy value`, the value copied isn't the one I'm hovering on, but its actual `value` - I guess that makes sense? 
image

- Sorts don't affect the funnel chart viz(?) - I think I would expect it to flip if I sort `Order status` in descending alphabetical order? 🤷

**Comment by joaoviana** [comment]
**WIP, will edit this as I go**

(non-blocking)
`getCreateEventProperties` in `SavedChartService` should include `funnel`


Tested user scenarios
Saving & Dashboard interaction
1. Added a saved funnel chart to dashboard and the filter in dashboards was applied correctly
2. Then went to `explore from here` and that also worked
3. Turned off the filter and saw the visualization change
Saving `explore from here` and directly into a dashboard
- all worked well
Image exports
- Exports in saved explorer
- Dashboard exports
- Scheduled deliveries
View underlying data 
- works as expected for both `data orientation` types

**Comment by magnew** [comment]
> Not relevant for this PR but when using columns as the steps - not being able to easily re-order the columns in the table or in the chart config could be quite annoying

@hamzahc1 Agreed. @joaoviana, you also mentioned sorting. We can re-visit this, but at the moment it is set up to always sort in descending order. This is actually the default for the echarts funnel and I think it's how many of them work. We could add sorting options, but this is fairly easy to explain/justify.

**Review by joaoviana — COMMENTED** [review]
Great work @magnew ! 
LMK what you think of the comments I've left

Also

(non-blocking)
`getCreateEventProperties` in `SavedChartService` should include `funnel`

File: packages/frontend/src/components/FunnelChart/index.tsx [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+        />
+    </div>
+);
+
+const LoadingChart = () => (
+    <div style={{ height: '100%', width: '100%', padding: '50px 0' }}>
+        <SuboptimalState
+            title="Loading chart"
+            loading
+            className="loading_chart"
```
Comment (IrakliJani) [inline packages/frontend/src/components/FunnelChart/index.tsx:32]: I believe this is for scheduled deliveries?

File: packages/frontend/src/hooks/useFunnelChartConfig.ts [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    tableCalculationsMetadata,
+) => {
+    const [fieldId, setFieldId] = useState(funnelChartConfig?.fieldId ?? null);
+
+    const [dataInput, setDataInput] = useState(
+        funnelChartConfig?.dataInput ?? FunnelChartDataInput.ROW,
+    );
+
+    // The value at the top of the funnel.
+    const maxValue = useRef(0);
```
Comment (IrakliJani) [inline packages/frontend/src/hooks/useFunnelChartConfig.ts:58]: why do we have this?

**Review by IrakliJani — COMMENTED** [review]
code looks solid! I added a few questions but did not have the chance to test the UI.

File: packages/frontend/src/components/Explorer/VisualizationCardOptions/index.tsx [hunk 3]
```diff
@@ -163,6 +165,11 @@ const VisualizationCardOptions: FC = memo(() => {
                     text: 'Pie chart',
                     icon: <MantineIcon icon={IconChartPie} color="gray" />,
                 };
+            case ChartType.FUNNEL:
+                return {
+                    text: 'Funnel chart',
+                    icon: <MantineIcon icon={IconFilter} color="gray" />,
```
[the remark this answers is not in the document]
Comment (magnew) [reply packages/frontend/src/components/Explorer/VisualizationCardOptions/index.tsx:171]: Haha, I was looking for that one and didn't see it. Thanks, will update.

File: packages/frontend/src/components/FunnelChart/FunnelChartContextMenu.tsx
Comment (magnew) [reply packages/frontend/src/components/FunnelChart/FunnelChartContextMenu.tsx:1]: I won't argue 😄. It was a place I was thinking about reducing repetition, but it's true that they are diverging mroe.

File: packages/frontend/src/components/FunnelChart/index.tsx [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+} from '../../hooks/echarts/useEchartsFunnelConfig';
+import { useApp } from '../../providers/AppProvider';
+import SuboptimalState from '../common/SuboptimalState/SuboptimalState';
+import { useVisualizationContext } from '../LightdashVisualization/VisualizationProvider';
+import FunnelChartContextMenu, {
+    type FunnelChartContextMenuProps,
+} from './FunnelChartContextMenu';
+
+const EmptyChart = () => (
+    <div style={{ height: '100%', width: '100%', padding: '50px 0' }}>
```
[the remark this answers is not in the document]
Comment (magnew) [reply packages/frontend/src/components/FunnelChart/index.tsx:18]: Yup, sure.

File: packages/frontend/src/components/FunnelChart/index.tsx (hunk 1 above)
Comment (magnew) [reply packages/frontend/src/components/FunnelChart/index.tsx:32]: The classname? Does that mean it should come out?

File: packages/frontend/src/hooks/useFunnelChartConfig.ts (hunk 2 above)
Comment (magnew) [reply packages/frontend/src/hooks/useFunnelChartConfig.ts:58]: This is the value that the funnel 'starts' with. Percentages in later stages are based on this number. I was surprised how bad a job echarts does here -- we have to calculate percentages ourselves to get it right. 

I will make the comment more clear

**Comment by magnew** [comment]
> Great work @magnew ! LMK what you think of the comments I've left
> 
> Also
> 
> (non-blocking) `getCreateEventProperties` in `SavedChartService` should include `funnel`

Good call. I missed that.

File: packages/frontend/src/components/Explorer/VisualizationCardOptions/index.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (magnew) [reply packages/frontend/src/components/Explorer/VisualizationCardOptions/index.tsx:171]: Ok, part of why I didn't find that is it's not in our tabler version. I'm going to leave it as the funnel for now and file a ticket to update tabler.

[the remark this answers is not in the document]
Comment (magnew) [reply packages/frontend/src/components/Explorer/VisualizationCardOptions/index.tsx:171]: https://github.com/lightdash/lightdash/issues/10475
https://github.com/lightdash/lightdash/issues/10476

File: packages/frontend/src/hooks/useFunnelChartConfig.ts [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+                    name: rowValues[0].formatted,
+                    value: dataValue,
+                    meta: {
+                        value: rowValues[fieldIndex],
+                        rows: [row],
+                    },
+                };
+            });
+        } else {
+            return allNumericFieldIds.reduce((acc, id) => {
```
[the remark this answers is not in the document]
Comment (magnew) [reply packages/frontend/src/hooks/useFunnelChartConfig.ts:137]: Done, although I see these as basically the exact same amount of code :D

**Comment by magnew** [comment]
> Early feedback (you can record these for future tickets, non-blocking)
> 
> * When there's no metric selected I would like to see an empty state
> 
> image
> * When I have data orientation as `columns` and I hover on the chart and click `copy value`, the value copied isn't the one I'm hovering on, but its actual `value` - I guess that makes sense?
> 
> image
> * Sorts don't affect the funnel chart viz(?) - I think I would expect it to flip if I sort `Order status` in descending alphabetical order? 🤷

Fixed the empty state. I think the copy issue is correct? You get the value instead of the label. We already talked about sorting.

File: packages/frontend/src/components/FunnelChart/index.tsx (hunk 1 above)
Comment (joaoviana) [reply packages/frontend/src/components/FunnelChart/index.tsx:32]: Correct, this is for scheduled deliveries. We should keep it so we can keep track of the loading spinners

File: packages/frontend/src/components/LightdashVisualization/VisualizationConfigFunnel.tsx [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    colorPalette,
+    children,
+    tableCalculationsMetadata,
+}) => {
+    const { dimensions, numericFields } = useMemo(() => {
+        const metrics = getMetricsFromItemsMap(itemsMap ?? {}, isNumericItem);
+        const tableCalculations = getTableCalculationsFromItemsMap(itemsMap);
+        return {
+            dimensions: getDimensionsFromItemsMap(itemsMap ?? {}),
+            numericFields: { ...metrics, ...tableCalculations },
```
[the remark this answers is not in the document]
Comment (magnew) [reply packages/frontend/src/components/LightdashVisualization/VisualizationConfigFunnel.tsx:55]: Added a numeric fitler here. I'll look into pie charts.

**Comment by magnew** [comment]
@joaoviana and @IrakliJani I think I have addressed all of the feedback so far. Thank you for the good reviews!

**Review by joaoviana — APPROVED** [review]
Tested Table calculations
- Created a number
- added to the funnel chart
- the changed its type to `string` 
- and it de-selected that tc and found a metric and selected it - great!

Approving as the improvements and other considerations will be done in different PRs later

**Pull request merged**